pkg/util/nsenter/nsenter.go

```diff
@@ -0,0 +1,139 @@
+	HostProcMountsPath = "/rootfs/proc/1/mounts"
+	// HostProcMountinfoPath is the default mount info path for rootfs
+	HostProcMountinfoPath = "/rootfs/proc/1/mountinfo"
+	// HostProcMountNsPath is the default mount namespace for rootfs
+	HostProcMountNsPath = "/rootfs/proc/1/ns/mnt"
+	// NsenterPath is the default nsenter command
+	NsenterPath = "nsenter"
+)
+
+// Nsenter is part of experimental support for running the kubelet
+// in a container.
+//
+// Nsenter requires:
+//
+// 1.  Docker >= 1.6 due to the dependency on the slave propagation mode
+//     of the bind-mount of the kubelet root directory in the container.
+//     Docker 1.5 used a private propagation mode for bind-mounts, so mounts
+//     performed in the host's mount namespace do not propagate out to the
+//     bind-mount in this docker version.
+// 2.  The host's root filesystem must be available at /rootfs
+// 3.  The nsenter binary must be on the Kubelet process' PATH in the container's
+//     filesystem.
+// 4.  The Kubelet process must have CAP_SYS_ADMIN (required by nsenter); at
+//     the present, this effectively means that the kubelet is running in a
+//     privileged container.
+// 5.  The volume path used by the Kubelet must be the same inside and outside
+//     the container and be writable by the container (to initialize volume)
+//     contents. TODO: remove this requirement.
+// 6.  The host image must have "mount", "findmnt", "umount", "stat", "touch",
+//     "mkdir", "ls", "sh" and "chmod" binaries in /bin, /usr/sbin, or /usr/bin
+// 7.  The host image should have systemd-run in /bin, /usr/sbin, or /usr/bin
+// For more information about mount propagation modes, see:
+//   https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt
+type Nsenter struct {
+	// a map of commands to their paths on the host filesystem
+	paths map[string]string
+}
+
+// NewNsenter constructs a new instance of Nsenter
+func NewNsenter() *Nsenter {
+	ne := &Nsenter{
+		paths: map[string]string{
+			"mount":       "",
+			"findmnt":     "",
+			"umount":      "",
+			"systemd-run": "",
+			"stat":        "",
+			"touch":       "",
+			"mkdir":       "",
+			"ls":          "",
+			"sh":          "",
+			"chmod":       "",
+		},
+	}
+	// search for the required commands in other locations besides /usr/bin
+	for binary := range ne.paths {
+		// default to root
+		ne.paths[binary] = filepath.Join("/", binary)
+		for _, path := range []string{"/bin", "/usr/sbin", "/usr/bin"} {
+			binPath := filepath.Join(path, binary)
+			if _, err := os.Stat(filepath.Join(hostRootFsPath, binPath)); err != nil {
+				continue
+			}
+			ne.paths[binary] = binPath
+			break
+		}
+		// TODO: error, so that the kubelet can stop if the paths don't exist
+		// (don't forget that systemd-run is optional)
+	}
+	return ne
+}
+
+// Exec executes nsenter commands in HostProcMountNsPath mount namespace
+func (ne *Nsenter) Exec(args ...string) exec.Cmd {
```

Comment:
This is most probably the only function that should be public. All nsenter users should use this and all other functions should be probably private (or removed). It should check that `args[0]` has full path (i.e. starts with '/') and automatically find the full path with `AbsHostPath` if it's not.

Btw, I'd prefer to have `(cmd string, args []string)` as argument so it is more obvious.

PS: `makeNsenterArgs` needs `SupportSystemd`, so that could be public too, but it's quite ugly...

Reply:
`AbsHostPath` should also be public due to the usage in `makeNsenterArgs`, which will be wrapped `/bin/mount` in `systemd-run` if supporting systemd.

Also ugly, but no other choice.